karapuzio/assignment2/BinaryTree.java

```diff
@@ -0,0 +1,172 @@
+            return;
+        }
+        if (temp == null) {
+            throw new BinaryTreeInitializationException("Check the parameters");
+        }
+        Node newChild = new Node();
+        newChild.value = value;
+        if (direction.charAt(direction.length()-1) == 'L'){
+            temp.left = newChild;
+        }
+        else{
+            temp.right = newChild;
+        }
+    }
+
+    /**
+     * Deep-first searching for save the traversal of tree.
+     * @param temp is current Node.
+     * @param traversal is support traversal.
+     * @param pos is the index in out tree.
+     */
+    private void dfs(Node temp, Map<T, Integer>traversal, int pos){
+        if (temp == null){
+            return;
+        }
+        traversal.put(temp.value, pos);
+        dfs(temp.left, traversal, pos*2);
+        dfs(temp.right, traversal, pos*2 + 1);
+    }
+
+    /**
+     * Method for call dfs with proper parameters.
+     */
+    public void treeTraversal(){
+        int position = 1;
+        Node temp = head;
+        dfs(temp, traversal, position);
+    }
+
+    /**
+     * The method to get all ancestors for the concrete Node from bottom to top.
+     * @param value is the key.
+     * @throws BinarySearchElementException
+     * return the list of all ancestors
+     */
+    public List<T> getAllAncestors(T value) throws BinarySearchElementException {
+        int pos = traversal.get(value);
+        if (pos == 0){
+            throw new BinarySearchElementException("There is sno this element in BT");
+        }
+        ArrayList<T> path = new ArrayList<>();
+        Node cur = findRoute(pos, path);
+        path.remove(path.size()-1);
+        System.out.println(path);
+        return path;
+    }
+
+    /**
+     * Find route by recursion steps.
+     * @param pos the current position of node in tree.
+     * @param route the route that we are searching.
+     * @return the node in the position pos.
+     */
+    private Node findRoute(int pos, List<T> route){
+        if (pos == 1){
+            route.add(head.value);
+            System.out.println(head.value);
+            return head;
+        }
+        if (pos % 2 == 0) {
+            Node temp = findRoute(pos/2, route).left;
+            route.add(temp.value);
+            System.out.println(temp.value);
+            return temp;
+        }
+        else{
+            Node temp = findRoute(pos/2, route).left;
+            route.add(temp.value);
+            System.out.println(temp.value);
+            return temp;
+        }
+    }
+
+    /**
+     * Method to get common ancestors.
+     * @param value1 is the first key.
+     * @param value2 is the second key.
+     * @return the least common ancestor.
+     * @throws BinarySearchElementException
+     */
+    public T getCommonAncestors(int value1, int value2) throws BinarySearchElementException {
+        int pos1 = traversal.get(value1);
+        int pos2 = traversal.get(value2);
+        if (pos1 == 0 || pos2 == 0){
```

Comment:
This won't be 0 if the node is not present in the TreeMap. It will return null. What will happen then?
Can you add a test that would catch it?

Reply:
I rewrite it with the checking if the key exist.

Reply:
And I added the test.